design-proposals/klip-27-enhanced-udf-configuration-options.md

```diff
@@ -0,0 +1,111 @@
+**Discussion**: [GitHub PR](https://github.com/confluentinc/ksql/pull/5269)
+
+**tl;dr:** This KLIP suggests to introduce a new configuration option in order to define UDF properties, that can be shared among a group / family of (custom) UDFs which logically belong together. At the same time, support for the `_global_` scope of configuration settings should get dropped.
+
+## Motivation and background
+
+Currently, ksqlDB supports two different ways to configure UDFs based on entries defined in `ksql-server.properties`:
+
+* either the configuration properties are directly bound to a single, specific UDF class and applicable to potentially existing overloaded methods - e.g. `ksql.functions.<myudfname>.<mysetting>=foo`
+
+* or we can have properties defined in a special scope, making them globally accessible - e.g. `ksql.functions._global_.<mysetting>=1234`
+
+The latter approach means that properties are readable by any(!) other UDF that gets loaded during bootstrap. While this may or may not be an issue depending on the type of properties, in general, it can be considered less than ideal. For instance, this leads to security concerns for sensitive configuration properties such as secrets / tokens / passwords, which is why the `_global_` scope should get dropped.
+
+## What is in scope
+
+This KLIP aims to allow for multiple, yet specifically defined UDFs, to share configuration properties other than "misusing" the global properties scope.
+
+## What is not in scope
+
+The scope of this KLIP is not a complete rewrite of the current configuration mechanisms, but a modified implementation that suppports the sharing of configuration properties between specific (custom) UDFs.
+
+## Value/Return
+
+The main value of this KLIP lies in the fact that users can easily define certain configuration properties that are shared by N >= 2 UDFs. This can be done without specifying such properties multiple times, once for each UDF in need, or without "misusing the global scope", both of which can be considered workarounds at best.
+
+## Public APIS
+
+There should neither be public API changes nor any KSQL query language modifications necessary. Ideally, the chosen implementation allows for existing configurations to be upgraded / migrated to newer ksqlDB versions without changes for configurations tied to specific UDFs.
+
+**Note however, that for workloads which make use of the `_global_` scope it will lead to a breaking change since support for this  gets dropped.**
+
+## Design
+
+A simple approach is to extend the function meta-data and add a _groups_ option to the `@UdfDecription` annotation, which takes a list of strings. Each string entry in _groups_ explicitly defines a name used to logically group functions for which to share configuration properties.
+
+Given the following example:
+
+```java
+@UdfDescription(
+    name = "someudf",
+    description = "...",
+    author = "...",
+    version = "...",
+    groups = {"shareit"/*,"..."*/})
```

Comment:
Do we feel there is benefit from allowing a Udf to be a member of multiple groups?  Can we think of usecases where this would be useful.  I guess we could have a company group, and a department group...

Reply:
I'm fine to go with a single group string.

Reply:
@hpgrahsl I should have really deleted this comment. It was more me thinking it through. I think there could be a use for multiple groups.  If it doesn't complicate the code that much, why not.

@agavra what do you reckon?

Reply:
Yeah, I think it makes sense to keep it as an array, very little overhead and it might be hard to change it later if we decide there is a use case for a multiple groups

Reply:
thx. for your quick answer!